### ISSUE TITLE
[Discussion] Parameter types

### ISSUE THREAD

**Issue by fieldOfView** [z25/pyZOCP#64]
Currently ZOCP has "convenience methods" for int, float, percent, bool, string, vec2f, vec3f and vec4f. Perhaps it is time to reconsider this list.

Out of the current 8, percent strikes me as a rather arbitrary choice; a percentage is just a float that (normally!) ranges between 0.0 and 100.0. The only place where the percent type is relevant at the moment is in Urwid, where it gets a different editor than a regular float.

Proposals for new types:
mat3x3:
mat4x4:
    eg transform matrices (list of 9 resp 16 floats)
object:
    basically a dict
binary:
    basically the same as string, without any charset conversion/limitations.
variant:
    don't care what type of data is being send, up to the receiver to decode (anything that can be serialised) NB: in the current implementation, because there is no type checking in ZOCP, all parameters behave like a variant type...

I have been mulling over adding color to the list, but a color can be considered a vec3f or vec4f the same way a percentage can be considered a float value.

**Comment by sphaero** [comment]
There is a difference between the data signature and the data type. A cartesian vector is 3 floats. A color is also 3 floats. So their signature is both `fff` however their type_hint is vector and color. 
Currently we have the method:
```
def _register_param(self, name, value, type_hint, access='r', min=None, max=None, step=None):
```
In my capability_refactor branch I proposed the following:
```
def _register_param(self, name, value, type_hint, signature, access='r', min=None, max=None, step=None):
```
So you can explicitly set the signature and still hint its type. The type hint is very usefull for generating UI widgets.

From there all suggestions are fine. Frequently used datatypes can be catered for using convenience methods.

I'm still doubting whether to follow pythons data type hinting:
https://docs.python.org/2/library/struct.html#format-characters

Or zeromq's: (https://github.com/zeromq/czmq#toc4-3930)
```
//  Send a 'picture' message to the socket (or actor). The picture is a   
//  string that defines the type of each frame. This makes it easy to send
//  a complex multiframe message in one call. The picture can contain any 
//  of these characters, each corresponding to one or two arguments:      
//                                                                        
//      i = int (signed)                                                  
//      1 = uint8_t                                                       
//      2 = uint16_t                                                      
//      4 = uint32_t                                                      
//      8 = uint64_t                                                      
//      b = byte *, size_t (2 arguments)                                  
//      c = zchunk_t *                                                    
//      f = zframe_t *                                                    
//      h = zhashx_t *                                                    
//      p = void * (sends the pointer value, only meaningful over inproc) 
//      m = zmsg_t * (sends all frames in the zmsg)                       
//      z = sends zero-sized frame (0 arguments)                          
//      u = uint (deprecated)                                             
//                                                                        
//  Note that s, b, c, and f are encoded the same way and the choice is   
//  offered as a convenience to the sender, which may or may not already  
//  have data in a zchunk or zframe. Does not change or take ownership of 
//  any arguments. Returns 0 if successful, -1 if sending failed for any  
//  reason.                               
```
